runtime/src/gpu/nvidia/Makefile.share

```diff
@@ -21,7 +21,17 @@ SRCS = $(GPU_SRCS)
 
 GPU_OBJS = $(addprefix $(GPU_OBJDIR)/,$(addsuffix .o,$(basename $(GPU_SRCS))))
 
-RUNTIME_CXXFLAGS += -x cuda -Wno-unknown-cuda-version
+# 1. our bundled clang (15 as of today) doesn't support warp-reductions which
+# came in with 8.x. So, we are limiting our runtime binary to 7.5. This seems to
+# work fine even if the user code is compiled with >7.5.
+# 2. clang-14 wants offload-arch to be passed multiple times and doesn't accept
+# a comma-separated list.
+RUNTIME_CXXFLAGS += -x cuda -Wno-unknown-cuda-version \
```

Comment:
Wait, isn't `--offload-arch` something that's being set here?:

https://github.com/search?q=repo%3Achapel-lang%2Fchapel+--offload-arch&type=code

If so, do we just need something that's more than sm_60? If so then it seems like `chplenv` script should check for that and error out if not.

Maybe I'm missing something though. Up until now we haven't actually compiled CUDA code into our runtime library (instead all of that is in functions in headers that we inline into your Chapel program). Is that no longer the case? (If so cool.)

Are we going to run into issues if this is misaligned with the gpu arches the user passes?

Reply:
> Wait, isn't --offload-arch something that's being set here?:

That's for the code that we compile.

> If so, do we just need something that's more than sm_60? If so then it seems like chplenv script should check for that and error out if not.

I am compiling this for all the arches we can and care about. >=`sm_80` doesn't seem to work, potentially because of clang-15 not being fully compatible with that. <`sm_60` we don't really care about.

> Maybe I'm missing something though. Up until now we haven't actually compiled CUDA code into our runtime library (instead all of that is in functions in headers that we inline into your Chapel program). Is that no longer the case? (If so cool.)

And, yes. This is the first time our runtime will have a GPU binary in it. It is not our code, but I am hoping that this can pave the way into us figuring out how to write our own runtime library for the device code. We are compiling a GPU binary, but the interface we use is host interface. However, the whole library is header-only, so the device code gets pulled into our runtime build directly.

> Are we going to run into issues if this is misaligned with the gpu arches the user passes?

I tried `sm_80` which is not there, and it worked. I think it all comes down to the fact that CUDA being backwards compatible.

Reply:
I guess my biggest worry is just how we'll know to keep this list up to date as new arches come about. I guess if backwards compatibility works that's less of a concern.

This is buried in a Makefile I wouldn't normally think about.

Any reason not to add `sm_80` (and others) here as well?

Reply:
> I guess my biggest worry is just how we'll know to keep this list up to date as new arches come about. I guess if backwards compatibility works that's less of a concern.

This is probably a part of the LLVM upgrade process somehow, as the supported sms are based on what clang allows.

> Any reason not to add sm_80 (and others) here as well?

Yeah: "I am compiling this for all the arches we can and care about. >=sm_80 doesn't seem to work, potentially because of clang-15 not being fully compatible with that". Specifically some warp-reduction primitive that's added in compute capability 8.x wasn't working with clang-15. The function is not in clang's CUDA headers at all. My theory is that clang-15 wasn't fully sm_80 compliant. I couldn't find a proof (or disproof) for that theory on a quick look, though.